packages/gui/src/components/nfts/provider/hooks/useNFTData.ts

```diff
@@ -0,0 +1,295 @@
+          nft,
+          isLoading: false,
+        });
+
+        events.emit('changed');
+      });
+
+      setLoaded((prevLoaded) => prevLoaded + page.length, abortSignal);
+
+      return nftIds;
+    },
+    [
+      pageSize /* immutable */,
+      getNFTs /* immutable */,
+      setLoaded /* immutable */,
+      nfts /* immutable */,
+      events /* immutable */,
+    ]
+  );
+
+  // immutable function
+  const fetchData = useCallback(
+    async (wallets: Wallet[], abortSignal: AbortSignal) => {
+      log('Fetching NFTs');
+      setCount(0, abortSignal);
+      setLoaded(0, abortSignal);
+
+      setErrorProcessing(undefined, abortSignal);
+      setIsLoadingProcessing(true, abortSignal);
+
+      const oldNFTIds = nfts.keys();
+      const newNFTIds = new Set<string>();
+
+      events.emit('changed');
+
+      async function processWallet(wallet: Wallet) {
+        log(`Processing wallet ${wallet.id}`);
+        if (abortSignal.aborted) {
+          return;
+        }
+
+        const { id: walletId } = wallet;
+        const { total } = await getNFTsCount({
+          walletIds: [walletId],
+        }).unwrap();
+
+        if (abortSignal.aborted) {
+          return;
+        }
+
+        setCount((prevCount) => prevCount + total, abortSignal);
+        const numPages = Math.ceil(total / pageSize);
+
+        const fetchLimited = (pageIndex: number) =>
+          add<string[]>(async () => {
+            const ids = await fetchNFTsPage(walletId, pageIndex, abortSignal);
+            ids.forEach((id) => newNFTIds.add(id));
+          });
+
+        const pageIndices = [];
+        for (let i = 0; i < numPages; i++) {
+          pageIndices.push(i);
+        }
+
+        await Promise.all(pageIndices.map(fetchLimited));
+      }
+
+      try {
+        await Promise.all(wallets.map(processWallet));
+      } catch (err) {
+        setErrorProcessing(err as Error, abortSignal);
+      } finally {
+        // remove old NFTs
+        if (!abortSignal.aborted) {
+          for (const nftId of oldNFTIds) {
+            if (!newNFTIds.has(nftId)) {
+              nfts.delete(nftId);
+            }
+          }
+        }
+
+        setIsLoadingProcessing(false, abortSignal);
+      }
+    },
+    [
+      nfts,
+      events,
+      getNFTsCount,
+      setCount,
+      setLoaded,
+      pageSize,
+      add,
+      fetchNFTsPage,
+      setErrorProcessing,
+      setIsLoadingProcessing,
+    ]
+  );
+
+  useEffect(() => {
+    // clear NFTs on fingerprint change
+    nfts.clear();
+  }, [nfts, fingerprint]);
+
+  const refetchData = useCallback(() => {
+    abortControllerRef.current.abort();
+    abortControllerRef.current = new AbortController();
+
+    return fetchData(nftWallets, abortControllerRef.current.signal);
+  }, [fetchData, nftWallets]);
+
+  useEffect(() => {
+    // abort previous fetch
+    if (fingerprint) {
+      let changed = false;
+      let initialLoadFinished = false;
+
+      refetchData().finally(() => {
```

Comment:
I haven't tested this, but if `refetchData()` calls `abort()`, does it throw an exception? If so, should that be caught and handled here?

Reply:
Same question would apply to the other callers of `refreshData()` below

Reply:
refetchData never emit any error because it is using setError